internal/jobworker/reexec/reexec.go

```diff
@@ -0,0 +1,161 @@
+type Job struct {
+	// ID is a unique identifier for the Job. The parent and child share the
+	// Job ID for each unique Job.
+	ID uuid.UUID
+	// Cmd is the arbitrary command to run as part of this Job.
+	Cmd Command
+}
+
+// Command represents a shell command.
+type Command struct {
+	// Name is the leading name of the command.
+	Name string
+	// Args are the arguments of the command.
+	Args []string
+}
+
+// Exec utilizes the piped data from the parent process to build and run a
+// arbitrary command on the host system.
+func Exec(ctx context.Context) (int, error) {
+	// Parent process has set /proc/self/fd/3 to the command pipe receiver.
+	cmdfd := os.NewFile(uintptr(3), "/proc/self/fd/3")
+	if cmdfd == nil {
+		return CommandFailure, ErrCommandPipeNotFound
+	}
+
+	// Parent process has set the /proc/self/fd/4 to the continue pipe receiver.
+	contfd := os.NewFile(uintptr(4), "/proc/self/fd/4")
+	if contfd == nil {
+		return CommandFailure, ErrContinuePipeNotFound
+	}
+
+	var buf bytes.Buffer
+	if _, err := buf.ReadFrom(cmdfd); err != nil {
+		return CommandFailure, fmt.Errorf("reexec read cmd in pipe; error: %w", err)
+	}
+	var job Job
+	if err := json.Unmarshal(buf.Bytes(), &job); err != nil {
+		return CommandFailure, fmt.Errorf("reexec unmarshal job; error: %w", err)
+	}
+
+	// Create log file for stdout and stderr output.
+	outfd, err := os.OpenFile(output.File(job.ID), os.O_CREATE|os.O_WRONLY, output.FileMode)
+	if err != nil {
+		return CommandFailure, fmt.Errorf("reexec open output file; error: %w", err)
+	}
+	defer func() {
+		if err := outfd.Close(); err != nil {
+			logger.Errorf("closing output fd; error: %s", err)
+		}
+	}()
+
+	// Build command to be run on host system.
+	cmd := exec.Command(job.Cmd.Name, job.Cmd.Args...)
+	cmd.Stdout = outfd
+	cmd.Stderr = outfd
+
+	// Wait for continue signal from parent process. This will be sent once
+	// process has been placed in the appropriate cgroup. If the continue signal
+	// is not retrieved within 10 seconds of waiting, cancel.
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+
+	if err := waitForContinue(ctx, contfd); err != nil {
+		return CommandFailure, fmt.Errorf("reexec wait for continue; error: %w", err)
+	}
+
+	if err := cmd.Start(); err != nil {
+		return CommandFailure, fmt.Errorf("start grandchild; error: %w", err)
+	}
+
+	err = cmd.Wait()
+	return exitCode(err), nil
+}
+
+func exitCode(err error) int {
+	if err == nil {
+		return CommandSuccess
+	}
+
+	exitError := new(exec.ExitError)
+	if errors.As(err, &exitError) {
+		status, ok := exitError.Sys().(syscall.WaitStatus)
+		if !ok {
+			return CommandFailure
+		}
+		return status.ExitStatus()
+	}
+	return CommandFailure
+}
+
+// waitForContinue waits for EOF to be received from fd. The parent process
+// will close fd's writer when this process may continue.
+func waitForContinue(ctx context.Context, fd io.Reader) error {
+	errc := make(chan error)
+	go func() {
```

Comment:
This goroutine may leak. If the context is canceled before an error can be written to `errc` then it will just hang forever.